### DISCUSSION TITLE
Native Ads

### DISCUSSION THREAD

**Discussion by yigit-serin** [admob-plus/admob-plus#311] [Fund Features]
Is there any way to add native ads support?

**Comment by ratson** [comment]
@yigit-serin I don't see a simple way to support it, as you cannot embed native view inside webview.

**Reply by CaturViriyananda** [reply]
hei, @ratson take a look at this [Cordova Native ads](https://developer.huawei.com/consumer/en/doc/development/HMS-Plugin-Guides-V1/native-ads-0000001050197495-V1) . might give you slight idea, maybe.

**Reply by ratson** [reply]
@CaturViriyananda It basically hijacked the DOM to overlay the ads, need lots of code and testing to make it right.
Would you want to fund this feature?

**Reply by CaturViriyananda** [reply]
i want to. but let me pump my traffic first :stuck_out_tongue_closed_eyes: 
![im poor right now](https://user-images.githubusercontent.com/32214264/115132792-9e048180-a02d-11eb-9c8d-b2292469e2a9.png)

**Comment by zlawson-ut** [comment]
@ratson I too have been on the hunt for a plugin that supports the native ad format so I can add inline ads for a better user-experience. I found [this](https://github.com/floatinghotpot/cordova-plugin-facebookads/blob/master/test/index.html) example of how it could potentially be implemented. I'd be happy to contribute to the development of it. Any resources or documentation you'd recommend starting with?

**Reply by ratson** [reply]
@zacharylawson You may reference the current BannerAd implement for creating the new NativeAd format, 
- https://github.com/admob-plus/admob-plus/blob/master/packages/cordova/ts/banner.ts
- https://github.com/admob-plus/admob-plus/blob/master/packages/cordova/src/ios/AMBBanner.swift
- https://github.com/admob-plus/admob-plus/blob/master/packages/cordova/src/android/ads/Banner.java

**Comment by creakosta** [comment]
@ratson ready pay for this $100

**Comment by yigit-serin** [comment]
I am currently minimal contribute to project, also I am ready to pay 100$ for this feature.

**Comment by eliadAfeka** [comment]
@ratson 150$

**Comment by zlawson-ut** [comment]
Ready to contribute $150 for this too

**Comment by eliadAfeka** [comment]
@ratson how much is missing to complete?

**Reply by zlawson-ut** [reply]
Based on [this discussion](https://github.com/admob-plus/admob-plus/discussions/314#discussion-3336690) it looks like we're 70% there currently.

**Reply by eliadAfeka** [reply]
@ratson I will complete the rest.

**Comment by Carlosps** [comment]
I will contribute $100 for this feature.

**Comment by mikend** [comment]
I will contribute the remaining 300

**Comment by zlawson-ut** [comment]
@ratson how much is missing to complete?

**Reply by sksk008** [reply]
it done @ratson @zacharylawson  https://github.com/admob-plus/admob-plus/discussions/314#discussion-3336690
i think its time to wait and watch

**Comment by ratson** [comment]
@creakosta @yigit-serin @eliadAfeka @zacharylawson @Carlosps @mikend 

I am going to start working on this feature, you may send your fund via
- https://github.com/sponsors/ratson?frequency=one-time
- https://www.paypal.com/paypalme/ratsonal

![](https://progress-bar.dev/60)

**Reply by creakosta** [reply]
Done

**Reply by Carlosps** [reply]
Done

**Reply by zlawson-ut** [reply]
Done

**Reply by EYALIN** [reply]
done (eliadAfeka)

**Reply by yigit-serin** [reply]
done (paid as leribol)

**Comment by ratson** [comment]
Just released `admob-plus-cordova@1.15.0` with Native Ad support.
The documentation lives in https://admob-plus.github.io/docs/cordova/ads/native
I would like to gather some feedback about the API to see if the design is make sense, while I am working on the code and docs about view customization.

**Reply by yigit-serin** [reply]
@ratson  thank you for the great job. I just tried, I think native ad must be attached to a div id, dom object etc. Currently native ad only attached to absolute location. (?)
You can check out Huawei's native ad implementation.
https://developer.huawei.com/consumer/en/doc/development/HMS-Plugin-Guides/native-ads-0000001050197495
These options might be good for native ads.

**Reply by ratson** [reply]
@yigit-serin Thank you for the feedback. The `show()` method could be called to multiple times to change size and position.

I think this way would enable user to bind the scroll or other events if needed.

A helper function to bind size and position to a DOM element will probably added once I complete the view customization feature.

Right now, it is open for the user space implementation. It would be awesome if anyone could figure that out and send a PR 😄

**Reply by zlawson-ut** [reply]
Excited to start using this, thanks @ratson!

I recently installed it via the instructions above but getting the following error when building a ionic/cordova app, any tips?

```
/platforms/ios/[...]/Plugins/admob-plus-cordova-native/AMNAdViewProvider.swift:3:36: error: cannot find type 'AMBNativeAdViewProvider' in scope
class AMNAdViewProvider: NSObject, AMBNativeAdViewProvider {
                                   ^~~~~~~~~~~~~~~~~~~~~~~
/platforms/ios/[...]/Plugins/admob-plus-cordova-native/AdMobNativePlugin.swift:6:19: error: type 'AMBPlugin' has no member 'registerNativeAdViewProviders'
AMBPlugin.registerNativeAdViewProviders(["default": AMNAdViewProvider()])

** ARCHIVE FAILED **

The following build commands failed:
	CompileSwift normal arm64 [...]/platforms/ios/[...]/Plugins/admob-plus-cordova-native/AMNAdViewProvider.swift
	CompileSwift normal arm64 [...]/platforms/ios/[...]/Plugins/admob-plus-cordova-native/AdMobNativePlugin.swift
	CompileSwiftSources normal arm64 com.apple.xcode.tools.swift.compiler
(3 failures)
```

**Reply by ollm** [reply]
> A helper function to bind size and position to a DOM element will probably added once I complete the view customization feature.

Some customizations that I have thought are these for the following examples:

**Example 1:**

Code:

``` javascript
  await ad.show({
    x: 0,
    y: 50,
    width: window.screen.width,
    height: 300,
    crop: {
       top: 50,
       right: 0,
       bottom: 0,
       left: 0,
    },
    elevations: [
       {
          elevation: 4, // dp
          x: 0, // Relative to ad
          y: -150, // Relative to ad
          width: window.screen.width,
          height: 200,
          color: '#000000',
       }
    ],
    foreground: {
       color: '#000000',
       opacity: 0,
    },
  })
```
Result:
![ad example 1](https://user-images.githubusercontent.com/18549191/120921870-3fbb5d00-c6c6-11eb-9e12-5720a66ac8e7.png)

**Example 2:**

Code:

``` javascript
  await ad.show({
    x: 0,
    y: 50,
    width: window.screen.width,
    height: 300,
    crop: {
       top: 50,
       right: 0,
       bottom: 0,
       left: 200,
    },
    elevations: [
       {
          elevation: 4, // dp
          x: 0, // Relative to ad
          y: -150, // Relative to ad
          width: window.screen.width,
          height: 200,
          color: '#000000',
       },
       {
          elevation: 8, // dp
          x: 0, // Relative to ad
          y: 0, // Relative to ad
          width: 200,
          height: 300,
          color: '#000000',
       }
    ],
    foreground: {
       color: '#000000',
       opacity: 0.5,
    },
  })
```

Result:
![ad example 2](https://user-images.githubusercontent.com/18549191/120923102-04705c80-c6cd-11eb-8364-bdde0ca7de3c.png)


Something that I had thought is to capture the view of the ad and drawit on a canvas element in the app so that it is really native, the clicks could be captured with the original view transparent above (With crop settings to avoid wrong clicks) or capture touch event from canvas and simulate in ad view (This most likely cannot be done), I do not know if all this is possible to do and if it would affect a lot performance (Although it could be optimized by reducing the frequency of drawing, stop it when the canvas is off the screen or when it is hidden (display: none)), one question I have is if this would violate AdMob policy.

What is your opinion on drawing the ad on a html canvas element?

Thanks for your plugin.

**Reply by Sayene** [reply]
@ratson 

> A helper function to bind size and position to a DOM element will probably added once I complete the view customization feature.

Yes, "inline" display of the native ad could be fairly easily done with sth like:

```js
async function _showNativeAd(elementId, adUnitId, height, width) {
  adUnitId = adUnitId || 'ca-app-pub-3940256099942544/2247696110';
  var nativeAd = new admob.NativeAd({adUnitId: adUnitId});
  var el = document.getElementById(elementId);

  nativeAd._show = function()  {
    var _width = width || window.screen.width;
    el.style.width = _width + 'px'; 
    el.style.height = height + 'px';
    var offsetTop = el.offsetTop - window.pageYOffset;
    return nativeAd.show({
        x: 0,
        y: offsetTop,
        width: _width,
        height: height
      })
  }
  await nativeAd.load();
  await nativeAd._show();
  document.addEventListener('scroll', nativeAd._show);
}
```
However, some issues with this approach are:
1. no scrolling when user scrolls over the ad (events are not passed, and if they were we'd have to handle them to ensure correct page scroll)
2. user experience - when scrolling webview and updating the ad by calling show():
- on Android: there is slight delay visible, depending on device etc. 
- on iOS there is rather bad lagg/jitter, especially when scrolling faster (looks much better in emulator than on real device)

I imagine we could expose subscription to touch/move events from the native ad and then handle them in js - emulating the webview scroll accordingly - however I'm not sure if this would be reliable and performant ... 
Maybe there is better way like syncing ad view and web view scroll events?
As alternative the idea to render the ad to the canvas element is interesting, but handling interactions and video could be a stopper there...

**Reply by Sayene** [reply]
and the funny part is that adMob ads are actually html rendered in web views ...

**Reply by elkhalifte** [reply]
i think its impossible to render native ads inside webview and show them responsively in scroll view . Google admob must done something for cordova developpers. 
i have taken a look into this [https://github.com/floatinghotpot/cordova-plugin-facebookads](url) but i dont know if that trick will cause any problem by the ad network since its not rendered natively.

**Reply by ollm** [reply]
I've been working on drawing the bannerAd on a html canvas and managed to get it to work decently (I have only developed it for android), the performance in the emulator is good (I have not tested it on a real device because I currently do not have any), where the refreshment is most noticeable is when the ads have animations, which go to stumbling blocks (Currently I have limited the update of the canvas every 250ms), the only problem with this is that I don't know if this violates AdMob policies.

**This is the repository that implements it**: https://github.com/ollm/admob-plus/tree/cordova-canvas-banner
**And here an example repository**:  https://github.com/ollm/admob-canvas-banner

If anyone can test it on a real device and comment how it works.

You can create an APK as follows:

```
git clone https://github.com/ollm/admob-canvas-banner.git
cordova build --debug android
cp platforms/android/app/build/outputs/apk/debug/app-debug.apk app-debug.apk
```

@ratson if the code seems good to you, I make the documentation and send a pull request.

**Screenshots:** 

<img src="https://user-images.githubusercontent.com/18549191/121580158-b8a31780-ca2c-11eb-86f7-65881b5d15d8.png" width="250"> <img src="https://user-images.githubusercontent.com/18549191/121580165-b93bae00-ca2c-11eb-98ec-9402d98a3fad.png" width="250"> <img src="https://user-images.githubusercontent.com/18549191/121580147-b640bd80-ca2c-11eb-90ef-e25252cedddf.png" width="250">

**Reply by ollm** [reply]
I have tried it on an old tablet that I have with the canvas refresh rate in 16.66ms, the animations of the ads are not completely fluid but they are not bad at all considering that the tablet is not very powerful.

**Videos:**
https://user-images.githubusercontent.com/18549191/121673372-18db9d00-cab1-11eb-8de9-ae42b279edee.mp4
https://user-images.githubusercontent.com/18549191/121673795-91daf480-cab1-11eb-8c37-d6f6fabec4d3.mp4

**Reply by ratson** [reply]
A new API `ad.showWith(document.getElementById('native-ad'))` has been added in `admob-plus-cordova@1.17.0`.

@Sayene 
> However, some issues with this approach are:
> 
>     1. no scrolling when user scrolls over the ad (events are not passed, and if they were we'd have to handle them to ensure correct page scroll)
> 
>     2. user experience - when scrolling webview and updating the ad by calling show():
> 
> 
>     * on Android: there is slight delay visible, depending on device etc.
> 
>     * on iOS there is rather bad lagg/jitter, especially when scrolling faster (looks much better in emulator than on real device)

I think this is the limitation for using NativeAd with webview. I don't see a way to avoid the overhead of scroll sync.

> As alternative the idea to render the ad to the canvas element is interesting, but handling interactions and video could be a stopper there...

I did think about render the Ad in HTML, but unsure if that is permitted and it won't work for video ads.

@ollm 
> I've been working on drawing the bannerAd on a html canvas and managed to get it to work decently (I have only developed it for android), the performance in the emulator is good (I have not tested it on a real device because I currently do not have any), where the refreshment is most noticeable is when the ads have animations, which go to stumbling blocks (Currently I have limited the update of the canvas every 250ms), the only problem with this is that I don't know if this violates AdMob policies.

That is an interesting approach, I am uncertain it fits the AdMob policies.
I think you create a separate plugin expose e.g. `CanvasBannerAd` for user to take the risk to use it if desired, I could link it in the doc.

**Reply by ollm** [reply]
> That is an interesting approach, I am uncertain it fits the AdMob policies.
> I think you create a separate plugin expose e.g. `CanvasBannerAd` for user to take the risk to use it if desired, I could link it in the doc.

It may be the best, when I can I will move the canvas banner to `CanvasBannerAd` class, currently I have it implemented like this:

**Normal banner**

``` javascript
await admob.start()

banner = await new admob.BannerAd({
	adUnitId: 'ca-app-pub-3940256099942544/6300978111',
})

await banner.show()
```
**Canvas banner**

``` javascript
await admob.start()

admob.BannerAd.config({
	canvasDrawInterval: 16.66, // 83.33,
	preciseDrawInterval: true,
});

banner = await new admob.BannerAd({
	adUnitId: 'ca-app-pub-3940256099942544/6300978111',
	canvas: document.querySelector('canvas'),
	autoDestroy: false,
})

await banner.show()
```
If when creating a banner you pass the `canvas` attribute, a banner will be created with the size of the canvas and it will be drawn on it, if the attribute is omitted it works as usual.

**Reply by elkhalifte** [reply]
@ollm wonderfull idea to implement native ads. i tested it on a real device (not a powerfull device) and i dont see any performance drop. the only thing matter is , would this approach go against admob policy.?

**Reply by elkhalifte** [reply]
and i dont think it will go against admob policy.

**Reply by ollm** [reply]
> and i dont think it will go against admob policy.

When it comes to Google you never know, sometimes they are strict.

> I think you create a separate plugin expose e.g. `CanvasBannerAd` for user to take the risk to use it if desired, I could link it in the doc.

@ratson I have made the changes and submitted a pull request.

**Reply by EYALIN** [reply]
@ratson is there any news regarding the native ad? thank you again for all your efforts

**Comment by EYALIN** [comment]
@ratson first, thank you for your great work!!
I started to try the native ad with the scroll update, but it looks like it killing the UI. it's a big burden for the app and makes debugging impossible.
is there another implementation solution for it? are they giving the option to put it inside the element?
another problem is that it goes over all app modals. 

I see that there is an option to attach it to view. can you show an example for it?

**Comment by mariusbolik** [comment]
Hi @ratson,

**I just donated $100** for your awesome work! Before I start trying around, I'd like to ask you if it'll be possible to use native ads with capacitor. This would be very important to me and I'm willing to donate more if native ads work with **capacitor**.

Little hint: also the [Cordova Google Maps Plugin](https://github.com/mapsplugin/cordova-plugin-googlemaps) works very well with scrolling, side menu and tabs. Maybe this information helps you :)

Menu | Tabs | Scrolling
--- | --- | ---
![ezgif-3-4213c5b3365c](https://user-images.githubusercontent.com/39862171/124453853-16a2f080-dd88-11eb-8e12-17c859cd9e25.gif) | ![ezgif-3-6fd18cb44049](https://user-images.githubusercontent.com/39862171/124453858-186cb400-dd88-11eb-9770-1fc77d6372a1.gif) | ![ezgif-3-2d4ec96c170f](https://user-images.githubusercontent.com/39862171/124453867-199de100-dd88-11eb-83e2-1f62c1fc7293.gif)

Best regards,
Marius

**Comment by EYALIN** [comment]
@ratson  is there any news regarding the native ad? we still cannot use it.

if you think that we need to add an extra donation for you, so feel free to ask, we all be more than happy to do so, in case we will have a working native ad.

we appreciate your hard work on this plugin. thanks !!

**Comment by jasonandress** [comment]
Any chance you can add standard Native Ad sizes? Small and Medium like provided in the following repo. I'm not as most people here a swift developer.

https://github.com/googleads/googleads-mobile-ios-native-templates/tree/master/nativetemplates
